src/utils/triton.h

```diff
@@ -0,0 +1,180 @@
+  if (t == 11)
+    return static_cast<TRITONSERVER_DataType>(1);
+  return static_cast<TRITONSERVER_DataType>(t + 2);
+}
+
+/**
+ * @brief Converts device_type_t to TRITONSERVER_MemoryType
+ */
+inline TRITONSERVER_MemoryType to_triton(device_type_t dev) {
+  switch (dev) {
+    case CPU:
+      return TRITONSERVER_MEMORY_CPU_PINNED;
+    case GPU:
+      return TRITONSERVER_MEMORY_GPU;
+    default:
+      throw std::invalid_argument("Unknown memory type");
+  }
+}
+
+class TritonInput {
+ public:
+  TritonInput(TRITONBACKEND_Input *handle) : handle_(handle) {
+    TRITONSERVER_DataType input_datatype;
+    const int64_t *input_shape;
+    uint32_t input_dims_count;
+    uint64_t input_byte_size;
+    uint32_t input_buffer_count;
+    const char *name;
+    TRITON_CALL_GUARD(TRITONBACKEND_InputProperties(handle_, &name, &input_datatype, &input_shape,
+                                                    &input_dims_count, &byte_size_, &buffer_cnt_));
+    meta_.name = std::string(name);
+    meta_.type = to_dali(input_datatype);
+    auto batch_size = input_shape[0];
+    auto sample_shape = TensorShape<>(input_shape + 1, input_shape + input_dims_count);
+    auto shape = TensorListShape<>::make_uniform(batch_size, sample_shape);
+    meta_.shape = shape;
+  }
+
+  IOMeta Meta() const {
+    return meta_;
+  }
+
+  size_t ByteSize() const {
+    return byte_size_;
+  }
+
+  uint32_t BufferCount() const {
+    return buffer_cnt_;
+  }
+
+  /**
+   * @brief Request an input buffer.
+   * @param idx Input index.
+   * @param device_type_t Preferred device type.
+   * @param device_id Preferred device id.
+   */
+  IBufferDescr GetBuffer(uint32_t idx, device_type_t device, int device_id) {
+    const void *data;
+    size_t size;
+    TRITONSERVER_MemoryType mem_type = to_triton(device);
+    int64_t mem_type_id = device_id;
+    TRITON_CALL_GUARD(
+        TRITONBACKEND_InputBuffer(handle_, idx, &data, &size, &mem_type, &mem_type_id));
+    IBufferDescr descr;
+    descr.device = to_dali(mem_type);
+    descr.device_id = mem_type_id;
+    descr.data = data;
+    descr.size = size;
+    return descr;
+  }
+
+ private:
+  TRITONBACKEND_Input *handle_;
+  IOMeta meta_;
+  size_t byte_size_;
+  uint32_t buffer_cnt_;
+};
+
+/**
+ * @brief RAII wrapper for a Triton request.
+ */
+class TritonRequest : public UniqueHandle<TRITONBACKEND_Request *, TritonRequest> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(TRITONBACKEND_Request *, TritonRequest)
+
+  static void DestroyHandle(TRITONBACKEND_Request *request) {
+    LOG_IF_ERROR(TRITONBACKEND_RequestRelease(
+                     request, TRITONSERVER_RequestReleaseFlag::TRITONSERVER_REQUEST_RELEASE_ALL),
+                 make_string("Failed releasing a request."));
+  }
+
+  uint32_t InputCount() {
+    uint32_t input_cnt;
+    TRITON_CALL_GUARD(TRITONBACKEND_RequestInputCount(handle_, &input_cnt));
+    return input_cnt;
+  }
+
+  TritonInput InputByIdx(uint32_t idx) {
```

Comment:
```suggestion
  TritonInput InputByIdx(uint32_t idx) const {
```
Even if the call modifies the request, it does not modify the handle/wrapper.

Reply:
A general remark about this class:
I think that adding C++ wrappers around Triton handles is a great idea, but they should be complete and that should be a separate PR. I will turn a blind eye on this here, but that mandates a follow-up IMO.
Also, there probably should be a TritonRequestView non-owning variant. We can discuss the pattern if you wish.

Reply:
This is how you can have an owning and non-owning variant and implement the wrapper functions only once:
```C++
template <class Actual>
class TritonRequestWrapper {
 public:
  uint32_t InputCount() const {
    uint32_t input_cnt;
    TRITON_CALL_GUARD(TRITONBACKEND_RequestInputCount(This().handle(), &input_cnt));
    return input_cnt;
  }

  TritonInput InputByIdx(uint32_t idx) const {
    TRITONBACKEND_Input *input;
    TRITON_CALL_GUARD(TRITONBACKEND_RequestInputByIndex(This().handle(), idx, &input));
    return TritonInput(input);
  }

  // add whatever other functions you need

 private:
  Actual &This() noexcept {
    return static_cast<Actual &>(*this);
  }

  const Actual &This() const noexcept {
    return static_cast<const Actual &>(*this);
  }
};


class TritonRequest : public UniqueHandle<TRITONBACKEND_Request *, TritonRequest>,
                      public TritonRequestWrapper<TritonRequest> {
 public:
  DALI_INHERIT_UNIQUE_HANDLE(TRITONBACKEND_Request *, TritonRequest)

  static void DestroyHandle(TRITONBACKEND_Request *request) {
    LOG_IF_ERROR(TRITONBACKEND_RequestRelease(
                     request, TRITONSERVER_RequestReleaseFlag::TRITONSERVER_REQUEST_RELEASE_ALL),
                 make_string("Failed releasing a request."));
  }
};

class TritonRequestView : public TritonRequestWrapper<TritonRequestView> {
 public:
  TritonRequestView() = default;

  TritonRequestView(TRITONBACKEND_Request *req) : handle_(req) {}  // NOLINT

  TRITONBACKEND_Request *handle() const noexcept {
    return handle_;
  }
 private:
  TRITONBACKEND_Request *handle_ = nullptr;
};
```

Reply:
I totally agree with the incompleteness of the TritonRequest, but I wanted to implement only the absolute minimum to not inflate this PR. Before I start adding new features, in the next PR I will implement all of the Triton types properly in a separate PR.

Reply:
That wrapper is a neat trick. Thanks!